src/blocks/weather/met_no.rs

```diff
@@ -0,0 +1,202 @@
+use super::*;
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "name", rename_all = "lowercase")]
+pub struct Config {
+    coordinates: Option<(String, String)>,
+    altitude: Option<String>,
+    lang: ApiLanguage,
+}
+
+#[derive(Deserialize)]
+struct LegendsResult {
+    desc_en: String,
+    desc_nb: String,
+    desc_nn: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub enum ApiLanguage {
+    #[serde(rename = "en")]
+    English,
+    #[serde(rename = "nn")]
+    NorwegianNynorsk,
+    #[serde(rename = "nb")]
+    NorwegianBokmaal,
+}
+
+impl Default for ApiLanguage {
+    fn default() -> Self {
+        ApiLanguage::English
+    }
+}
+
+impl From<&Option<&str>> for ApiLanguage {
+    fn from(o: &Option<&str>) -> Self {
+        match o.as_deref() {
+            Some(s) => match s {
+                "nn" => ApiLanguage::NorwegianNynorsk,
+                "nb" => ApiLanguage::NorwegianBokmaal,
+                _ => ApiLanguage::English,
+            },
+            None => ApiLanguage::English,
+        }
+    }
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastResponse {
+    properties: ForecastProperties,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastProperties {
+    timeseries: Vec<ForecastTimeStep>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastTimeStep {
+    data: ForecastData,
+    // time: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastData {
+    instant: ForecastModelInstant,
+    // next_12_hours: ForecastModelPeriod,
+    next_1_hours: Option<ForecastModelPeriod>,
+    // next_6_hours: ForecastModelPeriod,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastModelInstant {
+    details: ForecastTimeInstant,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastModelPeriod {
+    summary: ForecastSummary,
+}
+
+#[derive(Deserialize, Debug)]
+struct ForecastSummary {
+    symbol_code: String,
+}
+
+#[derive(Deserialize, Debug, Default)]
+struct ForecastTimeInstant {
+    air_temperature: Option<f64>,
+    wind_from_direction: Option<f64>,
+    wind_speed: Option<f64>,
+    relative_humidity: Option<f64>,
+}
+
+const LEGENDS_URL: &str = "https://api.met.no/weatherapi/weathericon/2.0/legends";
+const FORECAST_URL: &str = "https://api.met.no/weatherapi/locationforecast/2.0/compact";
+
+async fn translate(summary: &str, lang: &ApiLanguage) -> Result<String> {
+    let legend: HashMap<String, LegendsResult> = REQWEST_CLIENT
```

Comment:
I think this function can be executed just once

Reply:
Yes, indeed. But here I was kind of stumped by the modular architecture. Any global data must be initialized in the loop in weather.rs, and passed by reference to the get function. We could initialize a legends hashmap and then let the get function in met_no check if it is empty, and only then call get_legends to fetch the data and fill in the map. But how can we represent such a shared, mutable object in this async architecture?

 I tried to do this:

```rust
pub async fn run(config: toml::Value, mut api: CommonApi) -> Result<()> {
    ...
    let mut legends_store: HashMap<String, met_no::LegendsResult> = HashMap::new();

    loop {
        let data = api
            .recoverable(|| config.service.get(&mut legends_store, &location))
            .await?;
```


But that is disallowed by the compiler: captured variable cannot escape `FnMut` closure body.

Maybe an `Arc<Mutex<HashMap>>` would be allowed? But it seems bad to leek so much of the `met_no` specific logic into the weather.rs file.

Reply:
I do of course realise that I could just initialize it before the loop, if config is the MetNo variant. Maybe that is the best way to do it, even if it feels a bit dirty to initialize outside the met_no module.

Reply:
I think it's better to store config + state of each service in their own struct. You can look at `battery` or `keyboard_layout` block for examples.

Reply:
Ah, that is better. Fixed now (with the addition of Arc<Mutex>> for the config) in 7aa0069.